src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+        let mut outer_aggs = Vec::new();
+        let mut single_input_aggs_to_push = vec![(Vec::new(), Vec::new()); inputs.len()];
+        let mut no_input_aggs_to_push = Vec::new();
+
+        // classify the aggregations by how many inputs each aggregation references
+        for (agg_num, agg) in aggregates.iter().enumerate() {
+            let mut expr_inputs = old_join_mapper.lookup_inputs(&agg.expr);
+            if agg.func.outer_agg().is_none() {
+                // Aggregates that cannot be partially pushed down go in the
+                // outer reduce.
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if let AggregateFunc::Dummy = agg.func {
+                // Do not bother pushing dummy functions down.
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if agg.distinct && agg.func.ignores_duplicates() {
+                outer_aggs.push((agg_num, agg.clone()));
+            } else if let Some(expr_input) = expr_inputs.next() {
+                if expr_inputs.next().is_some() {
+                    // an aggregation involving more than one input
+                    outer_aggs.push((agg_num, agg.clone()));
+                } else {
+                    // an aggregation involving one input.
+                    let mut localized_agg = agg.clone();
+                    localized_agg.expr = old_join_mapper.map_expr_to_local(localized_agg.expr);
+                    if let MirRelationExpr::Reduce { group_key, .. } = &inputs[expr_input] {
+                        if localized_agg
+                            .expr
+                            .support()
+                            .iter()
+                            .any(|i| i >= &group_key.len())
+                        {
+                            // skip pushing down the aggregation if it aggregates a
+                            // partial aggregate to prevent infinite layers of
+                            // partial reduction pushdown
+                            outer_aggs.push((agg_num, agg.clone()));
+                            continue;
+                        }
+                    }
+                    single_input_aggs_to_push[expr_input]
+                        .0
+                        .push((agg_num, localized_agg));
+                }
+            } else {
+                // an aggregation involving no inputs
+                if agg.func.row_count_dependent() {
+                    no_input_aggs_to_push.push((agg_num, agg.clone()));
+                } else {
+                    // It is not worthwhile to push down a reduce whose result
+                    // only depends on whether or not an input is empty. such a
+                    // reduce should be simplified to `Distinct ()` + a map
+                    outer_aggs.push((agg_num, agg.clone()));
+                }
+            }
+        }
+
+        // Classify predicates by how many inputs the predicate references
```

Comment:
From a reader experience point of view, at this point I have no idea why we are doing this. Presumably this information feeds in to how we will plan things, but either I missed it, or it hasn't been stated yet.

Reply:
If this is about pushing down single-input predicates, should we just rely on predicate pushdown to do that?

Reply:
The answer is that it turns out that join equivalences can have single-input filters that then need to be pushed down. Philip's test caused that crash because join got into a state where it had a single-input filter in the equivalence.

I'm having trouble reproducing the crash at the moment, but you can easily get a join with a single-input filter in the equivalence like so:

```
materialize=> create table blah (a int, b int);
CREATE TABLE
materialize=> explain select a from blah, (select 1 as b) as foo where blah.b = foo.b;
            Optimized Plan             
---------------------------------------
 %0 =                                 +
 | Get materialize.public.blah (u35)  +
                                      +
 %1 =                                 +
 | Join %0 (= #1 1)                   +
 | | implementation = Differential %0 +
 | | demand = (#0)                    +
 | Project (#0)                       +
 
```

The LiteralLifting transform lifts the constant into the join equivalence class, and then the single input filter never gets pushed down. Upon looking at the predicate pushdown code, it looks like
1) pushing down an expr = literal equality out of a join equivalence has never been supported. 
2) pushing down an expr = expr equality out of a join equivalence seems to be broken for 11 months, and I guess we've just not noticed it.
https://github.com/MaterializeInc/materialize/blob/c6d230b5fa00e195aca12bb08845cb99cce86970/src/transform/src/predicate_pushdown.rs#L248

Let me cook up a PR to fix predicate pushdown.